src/topology/mod.rs

```diff
@@ -123,9 +123,14 @@ pub fn take_healthchecks(diff: &ConfigDiff, pieces: &mut Pieces) -> Vec<(String,
 }
 
 impl RunningTopology {
-    /// Returned future will finish once all current sources have finished.
+    /// Returned future will finish once all current sources have finished. If there are no
+    /// outputs, this will remain pending until re-called to allow for a provider to finish.
     pub fn sources_finished(&self) -> future::BoxFuture<'static, ()> {
-        self.shutdown_coordinator.shutdown_tripwire()
+        if self.outputs.is_empty() {
+            future::pending().boxed()
+        } else {
+            self.shutdown_coordinator.shutdown_tripwire()
+        }
```

Comment:
Right, this is to accommodate the scenario where there are no initial sources.

Loading the starting config synchronously is an option.

I think it'd require some work to the API to avoid it requiring an existing `RunningTopology`. I imagine we'd want to keep the API alive to be able to do things like status checks on a provider in the future.

I can do it that way if preferred, and move all topology until after the provider has initially and returned the first config.

Reply:
My two cents would be for this to be synchronous to easily allow use with `vector validate` and `vector test` as well.